File: packages/core/src/shadow/__tests__/ModelDictionary.test.ts
import { afterEach, describe, expect, it } from 'vitest';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

import { ModelDictionary, ModelEntry, ModelField } from '../ModelDictionary.js';

describe('ModelDictionary', () => {
  const tempDirs: string[] = [];

  afterEach(() => {
    for (const dir of tempDirs) {
      try {
        fs.rmSync(dir, { recursive: true, force: true });
      } catch {
        // ignore cleanup errors
      }
    }
    tempDirs.length = 0;
  });

  it('register should add model when no duplicate exists', async () => {
    const dict = new ModelDictionary();

    const entry: ModelEntry = {
      name: 'User',
      fields: [
        { name: 'id', type: 'string', required: true },
        { name: 'email', type: 'string', required: true },
        { name: 'name', type: 'string', required: false },
      ],
      relationships: [],
      source: 'packages/core/src/models/User.ts',
      tags: ['user', 'auth'],
    };

    const result = await dict.register(entry);
    expect(result.isDuplicate).toBe(false);
    expect(dict.getModelCount()).toBe(1);
  });

  it('checkDuplicate should detect exact name duplicate', async () => {
    const dict = new ModelDictionary();

    const entry1: ModelEntry = {
      name: 'User',
      fields: [
        { name: 'id', type: 'string', required: true },
        { name: 'email', type: 'string', required: true },
      ],
      relationships: [],
      source: 'models/User.ts',
      tags: ['user'],
    };

    const entry2: ModelEntry = {
      name: 'User',
      fields: [
        { name: 'id', type: 'number', required: true },
        { name: 'username', type: 'string', required: true },
      ],
      relationships: [],
      source: 'dto/UserDTO.ts',
      tags: ['user', 'dto'],
    };

    await dict.register(entry1);
    const result = dict.checkDuplicate(entry2);
    expect(result.isDuplicate).toBe(true);
    expect(result.similarModels.length).toBeGreaterThan(0);
    expect(result.similarModels[0].similarity).toBe(1.0);
  });

  it('checkDuplicate should detect structurally similar models', async () => {
    const dict = new ModelDictionary({ similarityThreshold: 0.3 });

    const fields: ModelField[] = [
      { name: 'id', type: 'string', required: true },
      { name: 'email', type: 'string', required: true },
      { name: 'name', type: 'string', required: false },
      { name: 'createdAt', type: 'number', required: true },
    ];

    const entry1: ModelEntry = {
      name: 'UserEntity',
      fields,
      relationships: [],
      source: 'entities/User.ts',
      tags: ['user', 'entity'],
    };

    const entry2: ModelEntry = {
      name: 'UserDTO',
      fields: [
        { name: 'id', type: 'string', required: true },
        { name: 'email', type: 'string', required: true },
        { name: 'name', type: 'string', required: false },
      ],
      relationships: [],
      source: 'dto/UserDTO.ts',
      tags: ['user', 'dto'],
    };

    await dict.register(entry1);
    const result = dict.checkDuplicate(entry2);
    expect(result.isDuplicate).toBe(true);
    expect(result.similarModels.length).toBeGreaterThan(0);
  });

  it('search should find matching models', async () => {
    const dict = new ModelDictionary();

    await dict.register({
      name: 'User',
      fields: [{ name: 'id', type: 'string', required: true }],
      relationships: [],
      source: 'models/User.ts',
      tags: ['user', 'auth'],
    });

    await dict.register({
      name: 'Order',
      fields: [{ name: 'id', type: 'string', required: true }],
      relationships: [],
      source: 'models/Order.ts',
      tags: ['order', 'commerce'],
    });

    const results = dict.search('user');
    expect(results.length).toBeGreaterThanOrEqual(1);
    expect(results[0].name).toBe('User');
  });

  it('recordRelationship should store entity-dto relationships', () => {
    const dict = new ModelDictionary();

    dict.recordRelationship('UserEntity', 'UserDTO', 'map');
    dict.recordRelationship('OrderEntity', 'OrderResponse', 'subset');
    dict.recordRelationship('UserEntity', 'UserDTO', 'map'); // duplicate

    const rels = dict.getRelationships();
    expect(rels).toHaveLength(2);
    expect(rels[0]).toEqual({ entity: 'UserEntity', dto: 'UserDTO', type: 'map' });
    expect(rels[1]).toEqual({ entity: 'OrderEntity', dto: 'OrderResponse', type: 'subset' });
  });

  it('saveToYaml and loadFromYaml should persist models', async () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'model-dict-'));
    tempDirs.push(tmpDir);

    const dictionaryPath = path.join(tmpDir, '.codeflow', 'registry', 'models.yaml');

    const dict1 = new ModelDictionary({ dictionaryPath });
    await dict1.register({
      name: 'User',
      fields: [
        { name: 'id', type: 'string', required: true },
        { name: 'email', type: 'string', required: true },
      ],
      relationships: [],
      source: 'models/User.ts',
      tags: ['user'],
    });
    dict1.recordRelationship('User', 'UserDTO', 'map');
    await dict1.saveToYaml();

    expect(fs.existsSync(dictionaryPath)).toBe(true);

    const dict2 = new ModelDictionary({ dictionaryPath });
    await dict2.loadFromYaml();
    expect(dict2.getModelCount()).toBe(1);
    expect(dict2.getModels()[0].name).toBe('User');
  });

  it('loadFromYaml should handle missing file gracefully', async () => {
    const dict = new ModelDictionary({ dictionaryPath: '/nonexistent/path/models.yaml' });
    await dict.loadFromYaml();
    expect(dict.getModelCount()).toBe(0);
  });
});

File: backend/internal/shadow/model_dictionary_test.go
package shadow

import (
	"os"
	"path/filepath"
	"testing"
)

func TestModelDictionaryRegister(t *testing.T) {
	dict := NewModelDictionary(nil)

	entry := ModelEntry{
		Name: "User",
		Fields: []ModelField{
			{Name: "id", Type: "string", Required: true},
			{Name: "email", Type: "string", Required: true},
			{Name: "name", Type: "string", Required: false},
		},
		Source: "models/User.ts",
		Tags:   []string{"user", "auth"},
	}

	result := dict.Register(entry)
	if result.IsDuplicate {
		t.Fatal("expected no duplicate on first register")
	}
	if dict.GetModelCount() != 1 {
		t.Fatalf("expected 1 model, got %d", dict.GetModelCount())
	}
}

func TestModelDictionaryCheckDuplicateExactName(t *testing.T) {
	dict := NewModelDictionary(nil)

	entry1 := ModelEntry{
		Name: "User",
		Fields: []ModelField{
			{Name: "id", Type: "string", Required: true},
			{Name: "email", Type: "string", Required: true},
		},
		Source: "models/User.ts",
		Tags:   []string{"user"},
	}
	dict.Register(entry1)

	entry2 := ModelEntry{
		Name: "User",
		Fields: []ModelField{
			{Name: "id", Type: "number", Required: true},
			{Name: "username", Type: "string", Required: true},
		},
		Source: "dto/UserDTO.ts",
		Tags:   []string{"user", "dto"},
	}

	result := dict.CheckDuplicate(entry2)
	if !result.IsDuplicate {
		t.Fatal("expected duplicate for same name")
	}
	if len(result.SimilarModels) == 0 {
		t.Fatal("expected at least one similar model")
	}
	if result.SimilarModels[0].Similarity != 1.0 {
		t.Fatalf("expected similarity 1.0, got %f", result.SimilarModels[0].Similarity)
	}
}

func TestModelDictionaryCheckDuplicateStructural(t *testing.T) {
	dict := NewModelDictionary(&ModelDictionaryConfig{SimilarityThreshold: 0.3})

	entry1 := ModelEntry{
		Name: "UserEntity",
		Fields: []ModelField{
			{Name: "id", Type: "string", Required: true},
			{Name: "email", Type: "string", Required: true},
			{Name: "name", Type: "string", Required: false},
			{Name: "createdAt", Type: "number", Required: true},
		},
		Source: "entities/User.ts",
		Tags:   []string{"user", "entity"},
	}
	dict.Register(entry1)

	entry2 := ModelEntry{
		Name: "UserDTO",
		Fields: []ModelField{
			{Name: "id", Type: "string", Required: true},
			{Name: "email", Type: "string", Required: true},
			{Name: "name", Type: "string", Required: false},
		},
		Source: "dto/UserDTO.ts",
		Tags:   []string{"user", "dto"},
	}

	result := dict.CheckDuplicate(entry2)
	if !result.IsDuplicate {
		t.Fatal("expected structural duplicate")
	}
}

func TestModelDictionarySearch(t *testing.T) {
	dict := NewModelDictionary(nil)

	dict.Register(ModelEntry{
		Name:   "User",
		Fields: []ModelField{{Name: "id", Type: "string", Required: true}},
		Source:  "models/User.ts",
		Tags:   []string{"user", "auth"},
	})
	dict.Register(ModelEntry{
		Name:   "Order",
		Fields: []ModelField{{Name: "id", Type: "string", Required: true}},
		Source:  "models/Order.ts",
		Tags:   []string{"order", "commerce"},
	})

	results := dict.Search("user")
	if len(results) < 1 {
		t.Fatalf("expected at least 1 result, got %d", len(results))
	}
	if results[0].Name != "User" {
		t.Fatalf("expected first result to be User, got %s", results[0].Name)
	}
}

func TestModelDictionaryRecordRelationship(t *testing.T) {
	dict := NewModelDictionary(nil)

	dict.RecordRelationship("UserEntity", "UserDTO", "map")
	dict.RecordRelationship("OrderEntity", "OrderResponse", "subset")
	dict.RecordRelationship("UserEntity", "UserDTO", "map") // duplicate

	rels := dict.GetRelationships()
	if len(rels) != 2 {
		t.Fatalf("expected 2 relationships, got %d", len(rels))
	}
	if rels[0].Entity != "UserEntity" || rels[0].DTO != "UserDTO" || rels[0].Type != "map" {
		t.Fatalf("unexpected first relationship: %+v", rels[0])
	}
}

func TestModelDictionarySaveAndLoad(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "model_dict_test")
	if err != nil {
		t.Fatalf("create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	dictPath := filepath.Join(tmpDir, ".codeflow", "registry", "models.yaml")

	d1 := NewModelDictionary(&ModelDictionaryConfig{DictionaryPath: dictPath})
	d1.Register(ModelEntry{
		Name: "User",
		Fields: []ModelField{
			{Name: "id", Type: "string", Required: true},
			{Name: "email", Type: "string", Required: true},
		},
		Source: "models/User.ts",
		Tags:   []string{"user"},
	})
	d1.RecordRelationship("User", "UserDTO", "map")

	if err := d1.SaveToYAML(); err != nil {
		t.Fatalf("save failed: %v", err)
	}

	d2 := NewModelDictionary(&ModelDictionaryConfig{DictionaryPath: dictPath})
	if err := d2.LoadFromYAML(); err != nil {
		t.Fatalf("load failed: %v", err)
	}

	if d2.GetModelCount() != 1 {
		t.Fatalf("expected 1 model, got %d", d2.GetModelCount())
	}
	models := d2.GetModels()
	if models[0].Name != "User" {
		t.Fatalf("unexpected model name: %s", models[0].Name)
	}
}

func TestModelDictionaryLoadMissingFile(t *testing.T) {
	d := NewModelDictionary(&ModelDictionaryConfig{DictionaryPath: "/nonexistent/path/models.yaml"})
	if err := d.LoadFromYAML(); err != nil {
		t.Fatalf("expected no error for missing file, got: %v", err)
	}
	if d.GetModelCount() != 0 {
		t.Fatalf("expected 0 models, got %d", d.GetModelCount())
	}
}
